pkg/metrics/cluster/node.go

```diff
@@ -176,59 +175,66 @@ func (n *node) run() {
 func (n *node) performClusterReshard(ctx context.Context, joining bool) error {
 	if n.ring == nil || n.lc == nil {
 		level.Info(n.log).Log("msg", "node disabled, not resharding")
-		return nil
-	}
-
-	if n.cfg.ReshardTimeout > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, n.cfg.ReshardTimeout)
-		defer cancel()
 	}
 
 	var (
 		rs  ring.ReplicationSet
 		err error
-
-		firstError error
 	)
 
+	// This cancel has to exist outside since it is used in the goroutine
+	// originally it was put in a if !joining block which would logically
+	// ensure it would be called, but the linter does not process those paths
+	// so instead its sprinkled around before the returns.
+	var cancel context.CancelFunc
+	if n.cfg.RefreshTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, n.cfg.RefreshTimeout)
```

Comment:
There should be a `defer cancel()` here so it can't leak.

Reply:
Yes I can do this now since we moved the goroutine to the reshard, beforehand this was needed because it would cancel before the goroutine was ran and always be cancelled.